### PR TITLE
Document breaking change for #46044

### LINKED ISSUE #51050: [.NET 6] [Source Compat Failure] DiscordChatExporter app doesn't build with default target 3.1 with .net6 SDK
**Application Name:**  DiscordChatExporter
**OS:** Windows 10 RS5
**CPU:** X64
**.NET Build Number:** 6.0.100-preview.1.21080.2

 **Verify Scenarios:**
1) App Target 3.1 + .NET Core SDK build 6.0.100-preview.1.21080.2: Fail
2) App Target 3.1 + .NET Core SDK build 5.0.102: Pass
3) App Re-Target 5.0 + .NET Core SDK build 5.0.102: Fail
4) App Re-Target 6.0 + .NET Core SDK build 6.0.100-preview.1.21080.2: Fail

**Repro steps**
Source code is targeting netcoreapp3.1(default Runtime) , global.json set to 6.0.100-preview.1.21080.2
1) Source Code check: [https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1272682](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1272682)
2) dotnet build

**Expected Result:**
Build succeeded

**Actual Result:**
```
C:\Users\valtunc\Desktop\Sources\DiscordChatExporter\DiscordChatExporter.Gui\ViewModels\Framework\DialogScreen.cs(13,44): error CS8601: Possible null reference assignment. [C:\Users\valtunc\Desktop\Sources\DiscordChatExporter\DiscordChatExporter.Gui\DiscordChatExporter.Gui_nlgh5as3_wpftmp.csproj]
Build FAILED.
```

**Findings:**
Since we are not changing default target(3.1), we think it should build successfully with net6 SDK(build succeeded with net5 SDK)
When we target to net5 or net6, we get issues related to nullable, to resolve nullable issues, we need to remove 
```
    <Nullable>enable</Nullable>
    <WarningsAsErrors>nullable</WarningsAsErrors>
```
inside Directory.Build.props file. 
But with the config above, we need to understand the different behaviors of default target 3.1 build with .NET 5 & .NET 6, and retarget 5 & 6 both failed.
```
1) App Target 3.1 + .NET Core SDK build 6.0.100-preview.1.21080.2: Fail
2) App Target 3.1 + .NET Core SDK build 5.0.102: Pass
3) App Re-Target 5.0 + .NET Core SDK build 5.0.102: Fail
4) App Re-Target 6.0 + .NET Core SDK build 6.0.100-preview.1.21080.2: Fail
```

@dotnet-actwx-bot  @dotnet/compat

### PR DESCRIPTION
**Pull request by cston**
Fixes #51050

### CODE REVIEW

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 6.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿## This document lists known breaking changes in Roslyn in C# 9.0 introduced with .NET 6.
```
Comment (Youssef1313) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 6.md:1]: Consider adding a link for this document in <https://github.com/dotnet/docs/blob/master/docs/csharp/whats-new/breaking-changes.md>.

File: docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md [hunk 2]
```diff
@@ -142,3 +142,15 @@ class C
 
 22. https://github.com/dotnet/roslyn/issues/49596 In *Visual Studio 2019 version 16.9* and greater, conversions from `sbyte` or `short` to `nuint` require explicit casts,
 and binary operations with `sbyte` or `short` and `nuint` arguments require explicit casts for one or both operands for `+`, `-`, `*`, `/`, `%`, `<`, `>`, `<=`, `>=`, `==`, `!=`, `|`, `&`, and `^`.
+
+23. https://github.com/dotnet/roslyn/issues/46044 In *Visual Studio 2019 version 16.9* and greater, a warning is reported when assigning `default` to, or when casting a possibly `null` value to a type parameter type that is not constrained to value types or reference types. To avoid the warning, the type can be annotated with `?`.
```
Comment (Youssef1313) [inline docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md:146]: I'm a little bit lost here. The title of this document is:

> This document lists known breaking changes in Roslyn in Visual Studio 2019 Update 1 and beyond compared to Visual Studio 2019.

I'm expecting this to be documented in the new file only. Am I missing something?

**Review by Youssef1313 — COMMENTED** [review]
A general question: From a time to time, there are PRs that introduces extra nullable warnings in specific contexts. But these were never documented as breaking changes. Why is this specific one special?

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 6.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+1. https://github.com/dotnet/roslyn/issues/46044 In C# 9.0 (Visual Studio 16.9), a warning is reported when assigning `default` to, or when casting a possibly `null` value to a type parameter type that is not constrained to value types or reference types. To avoid the warning, the type can be annotated with `?`.
```
Comment (jcouv) [inline docs/compilers/CSharp/Compiler Breaking Changes - DotNet 6.md:3]: >In C# 9.0 [](start = 49, length = 9)

I think this warning is now produced even for LangVer=8, right? If so we should just say "In Visual Studio 16.9, ..."

File: docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md (hunk 2 above)
Comment (jcouv) [reply docs/compilers/CSharp/Compiler Breaking Changes - post VS2019.md:146]: Indeed, it looks like we messed up with some recent breaking changes.
The latest file is https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%205.md and corresponds to VS 16.8 / .NET 5.
We should probably start a new file for post-.NET 5.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

**Comment by jcouv** [comment]
Marked as draft to make room in review queue.

File: docs/compilers/CSharp/Compiler Breaking Changes - DotNet 6.md (hunk 3 above)
Comment (cston) [reply docs/compilers/CSharp/Compiler Breaking Changes - DotNet 6.md:3]: > I think this warning is now produced even for LangVer=8, right?

No, the warnings are not reported with `/langversion:8`.

> So we could start a new document for post-DotNet 5.

I've renamed the document to "Compiler Breaking Changes - post DotNet 5.md" based on our discussion.

---
In reply to: [572483433](https://github.com/dotnet/roslyn/pull/51070#discussion_r572483433) [](ancestors = 572483433)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Pull request merged**